File: app/src/main/java/com/example/network_termproject/SelectFriend.java
package com.example.network_termproject;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.network_termproject.databinding.SelectFriendLayoutBinding;
import com.example.network_termproject.network.Client;
import com.example.network_termproject.network.NetData;
import com.example.network_termproject.recycler.SelectAdapter;

import java.util.ArrayList;
import java.util.Objects;

public class SelectFriend extends AppCompatActivity {

    private SelectFriendLayoutBinding binding;
    private ArrayList<String> users;
    private SelectAdapter selectAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = SelectFriendLayoutBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        users = new ArrayList<>();
        selectAdapter = new SelectAdapter(users);
        binding.selectRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        binding.selectRecyclerView.setAdapter(selectAdapter);
        binding.selectRecyclerView.setHasFixedSize(true);

        setSupportActionBar(binding.selectToolBar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);

        binding.selectOkButton.setOnClickListener(view -> {
            Intent intent = new Intent();
            intent.putExtra("selectedUserList", selectAdapter.getCheckedUsers());
            setResult(RESULT_OK, intent);
            finish();
        });

        requestList();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.home){
            Intent intent = new Intent();
            setResult(RESULT_CANCELED, intent);
            finish();
            Log.d("Options", "in");
        }
        return false;
    }

    private void requestList() {
        NetData.Builder builder = new NetData.Builder();
        Client.getInstance().setList(this::setList);
        Client.getInstance().write(builder.setType("requestList").setUserId(Client.getInstance().getId()).build());
    }

    private void setList(NetData serverData) {
        users.clear();
//        users.addAll(serverData.getList());
//        selectAdapter.notifyDataSetChanged();
        runOnUiThread(() -> {
            for(int i=0;i<serverData.getList().size();i++){
                users.add(serverData.getList().get(i));
                selectAdapter.notifyItemInserted(i);
            }
        });

    }

}

File: app/src/main/java/com/example/network_termproject/SelectFriend.kt
package com.example.network_termproject

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.network_termproject.network.Client
import com.example.network_termproject.network.NetData
import com.example.network_termproject.recycler.SelectAdapter
import kotlinx.android.synthetic.main.select_friend_layout.*
import java.nio.ByteBuffer
import java.util.*
import java.util.function.Consumer

class SelectFriend : AppCompatActivity() {

    private var users: ArrayList<String>? = null
    private var selectAdapter: SelectAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.select_friend_layout)

        users = ArrayList()
        selectAdapter = SelectAdapter(users!!)

        select_recyclerView.apply {
            layoutManager = LinearLayoutManager(this@SelectFriend)
            adapter = selectAdapter
            setHasFixedSize(true)
        }

        setSupportActionBar(select_toolBar)
        Objects.requireNonNull(supportActionBar)?.setDisplayHomeAsUpEnabled(true)
        select_okButton.setOnClickListener {
            val intent = Intent()
            intent.putExtra("selectedUserList", selectAdapter!!.checkedUsers)
            setResult(RESULT_OK, intent)
            finish()
        }
        requestList()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.home) {
            val intent = Intent()
            setResult(RESULT_CANCELED, intent)
            finish()
        }
        return false
    }

    private fun requestList() {
        val builder = NetData.Builder()
        Client.instance.setList(Consumer { serverData: NetData? -> setList(serverData!!) })
        val mainData = builder.setType("requestList").setUserId(Client.instance.id!!).build()

        val header = ByteBuffer.allocate(6)
        header.apply {
            putChar('s')
            putInt(mainData.data.toString().toByteArray().size)
            flip()
        }

        var buffer = ByteBuffer.allocate(6 + mainData.data.toString().toByteArray().size)
        buffer.apply {
            put(header)
            put(mainData.data.toString().toByteArray())
            flip()
        }
        Client.instance.write(buffer)
    }

    private fun setList(serverData: NetData) {
        users!!.clear()
        runOnUiThread {
            for (i in 0 until serverData.getList().size) {
                users!!.add(serverData.getList()[i])
                selectAdapter!!.notifyItemInserted(i)
            }
        }
    }
}